ark/spLDA/processing.py

```diff
@@ -184,3 +189,121 @@ def create_difference_matrices(cell_table, features, training=True, inference=Tr
+    """Computes various metrics for k-means clustering models to help determine an
+    appropriate number of topics for use in spatial-LDA analysis.
+
+    Args:
+        features (pandas.DataFrame):
+            A DataFrame of featurized cellular neighborhoods.  Specifically, this is one of the
+            outputs of :func:`~ark.spLDA.processing.featurize_cell_table`.
+        topics (list):
+            A list of integers corresponding to the different number of possible topics to
+            investigate.
+        num_boots (int):
+            The number of bootstrap samples to use when calculating the Gap-statistic.
+
+    Returns:
+        dict:
+
+        - A dictionary of dictionaries containing the corresponding metrics for each topic value
+        provided.
+
+    """
+    # Check inputs
+    if num_boots < 25:
+        raise ValueError("Number of bootstrap samples must be at least 25")
+    if min(topics) <= 2:
+        raise ValueError("Number of topics must be greater than 2")
+    if max(topics) >= features.shape[0] - 1:
+        raise ValueError("Number of topics must be less than {}".format(features.shape[0] - 1))
+
+    inertias, silhouettes, gap_stats, gap_sds, pct_vars = {}, {}, {}, {}, {}
+
+    # Compute the total sum of squared pairwise distances between all observations
+    total_ss = np.sum(pdist(features) ** 2) / features.shape[0]
+    for k in topics:
+        cluster_fit = KMeans(n_clusters=k).fit(features)
+        inertias[k] = cluster_fit.inertia_
+        silhouettes[k] = silhouette_score(features, cluster_fit.labels_, metric='euclidean')
+        gap_stats[k], gap_sds[k] = gap_stat(features, k, cluster_fit.inertia_, num_boots)
+        pct_vars[k] = (total_ss - cluster_fit.inertia_) / total_ss
+
+    stats = {"inertia": inertias, "silhouette": silhouettes,
+             "gap_stat": gap_stats, "gap_sds": gap_sds, "percent_var_exp": pct_vars}
+    return stats
+
+
+def fov_density(cell_table, total_pix=1024 ** 2):
+    """Computes cellular density metrics for each field of view to determine an appropriate
+    radius for the featurization step.
+
+    Args:
+        cell_table (dict):
+            A formatted cell table for use in spatial-LDA analysis. Specifically, this is the
+            output from :func:`~ark.spLDA.processing.format_cell_table`.
+        total_pix (int):
+            The total number of pixels in each field of view.
+
+    Returns:
+        dict:
+
+        - A dictionary containing the average cell size and the cellular density for each field
+        of view.  Cellular density is calculated by summing the total number of pixels occupied
+        by cells divided by the total number of pixels in each field of view.
+
+    """
+    density_stats = {
+        "average_area": {k: v["cell_size"].mean() for (k, v) in cell_table.items()},
+        "cellular_density": {
+            k: np.sum(v["cell_size"]) / total_pix for (k, v) in cell_table.items()}
```

Comment:
```suggestion
        "cellular_density": {
            k: np.sum(v["cell_size"]) / total_pix for (k, v) in cell_table.items()
        }
```

Reply:
I ended up having to change this whole part anyway because python got mad at me